src/ancillary/writer.rs

```diff
@@ -0,0 +1,217 @@
+///
+/// # Example
+/// ```no_run
+/// use tokio_seqpacket::UnixSeqpacket;
+/// use tokio_seqpacket::ancillary::AncillaryMessageWriter;
+/// use std::io::IoSlice;
+/// use std::os::fd::AsRawFd;
+///
+/// #[tokio::main]
+/// async fn main() -> std::io::Result<()> {
+///     let sock = UnixSeqpacket::connect("/tmp/sock").await?;
+///     let file = std::fs::File::create("/tmp/my-file")?;
+///
+///     let mut fds = [0; 8];
+///     let mut ancillary_buffer = [0; 128];
+///     let mut ancillary = AncillaryMessageWriter::new(&mut ancillary_buffer);
+///     ancillary.add_fds(&[&file])?;
+///
+///     let mut buf = [1; 8];
+///     let mut bufs = [IoSlice::new(&mut buf)];
+///     sock.send_vectored_with_ancillary(&mut bufs, &mut ancillary).await?;
+///
+///     Ok(())
+/// }
+/// ```
+#[derive(Debug)]
+pub struct AncillaryMessageWriter<'a> {
+	pub(crate) buffer: &'a mut [u8],
+	pub(crate) length: usize,
+}
+
+/// Failed to add a control message to a ancillary message buffer.
+pub struct AddControlMessageError(());
+
+impl<'a> AncillaryMessageWriter<'a> {
+	/// Create an ancillary data with the given buffer.
+	///
+	/// # Example
+	///
+	/// ```no_run
+	/// # #![allow(unused_mut)]
+	/// use tokio_seqpacket::ancillary::AncillaryMessageWriter;
+	/// let mut ancillary_buffer = [0; 128];
+	/// let mut ancillary = AncillaryMessageWriter::new(&mut ancillary_buffer);
+	/// ```
+	pub fn new(buffer: &'a mut [u8]) -> Self {
+		Self { buffer, length: 0 }
+	}
+
+	/// Returns the capacity of the buffer.
+	pub fn capacity(&self) -> usize {
+		self.buffer.len()
+	}
+
+	/// Returns `true` if the ancillary data is empty.
+	pub fn is_empty(&self) -> bool {
+		self.length == 0
+	}
+
+	/// Returns the number of used bytes.
+	pub fn len(&self) -> usize {
+		self.length
+	}
+
+	/// Add file descriptors to the ancillary data.
+	///
+	/// The function returns `Ok(())` if there was enough space in the buffer.
+	/// If there was not enough space then no file descriptors was appended.
+	///
+	/// This adds a single control message with level `SOL_SOCKET` and type `SCM_RIGHTS`.
+	///
+	/// # Example
+	///
+	/// ```no_run
+	/// use tokio_seqpacket::UnixSeqpacket;
+	/// use tokio_seqpacket::ancillary::AncillaryMessageWriter;
+	/// use std::os::unix::io::AsFd;
+	/// use std::io::IoSlice;
+	///
+	/// #[tokio::main]
+	/// async fn main() -> std::io::Result<()> {
+	///     let sock = UnixSeqpacket::connect("/tmp/sock").await?;
+	///     let file = std::fs::File::open("/my/file")?;
+	///
+	///     let mut ancillary_buffer = [0; 128];
+	///     let mut ancillary = AncillaryMessageWriter::new(&mut ancillary_buffer);
+	///     ancillary.add_fds(&[file.as_fd()]);
+	///
+	///     let buf = [1; 8];
+	///     let mut bufs = &mut [IoSlice::new(&buf)];
+	///     sock.send_vectored_with_ancillary(bufs, &mut ancillary).await?;
+	///     Ok(())
+	/// }
+	/// ```
+	pub fn add_fds<T>(&mut self, fds: &[T]) -> Result<(), AddControlMessageError>
```

Comment:
Discussed in person, but for completeness sake: the writer has to borrow a `&'a mut [u8]`, even if it can own file descriptors. So the `Writer` will never be '`static`, `Send` or `Sync`. Taking that into account, I see no benefit to support adding owned file descriptors to the writer.